README.md
# Magento 2 Kubernetes DevBox

[![Build Status](https://travis-ci.com/magento/magento2-kubernetes-devbox.svg?branch=master)](https://travis-ci.com/magento/magento2-kubernetes-devbox)
<!--[![MIT License](https://img.shields.io/badge/license-MIT-blue.svg)](./LICENSE)-->
<!--[![Semver](http://img.shields.io/SemVer/2.0.0.png?color=blue)](http://semver.org/spec/v2.0.0.html)-->
<!--[![Latest GitHub release](docs/images/release_badge.png)](https://github.com/paliarush/magento2-vagrant-for-developers/releases/latest)-->

 * [What You get](#what-you-get)
 * [How to install](#how-to-install)
   * [Requirements](#requirements)
   * [Installation steps](#installation-steps)
   * [Default credentials and settings](#default-credentials-and-settings)
   * [Getting updates and fixes](#getting-updates-and-fixes)
 * [Day-to-day development scenarios](#day-to-day-development-scenarios)
   * [Access Magento](#access-magento)
   * [Reinstall Magento](#reinstall-magento)
   * [Clear Magento cache](#clear-magento-cache)
   * [Switch between CE and EE](#switch-between-ce-and-ee)
   * [Sample data installation](#sample-data-installation)
   * [Basic data generation](#basic-data-generation)
   * [Use Magento CLI (bin/magento)](#use-magento-cli-binmagento)
   * [Debugging with XDebug](#debugging-with-xdebug)
   * [Connecting to MySQL DB](#connecting-to-mysql-db)
   * [View emails sent by Magento](#view-emails-sent-by-magento)
   * [Accessing PHP and other config files](#accessing-php-and-other-config-files)
   * [Upgrading Magento](#upgrading-magento)
   * [Multiple Magento instances](#multiple-magento-instances)
   * [Update Composer dependencies](#update-composer-dependencies)
   * [Running Magento tests](#running-magento-tests)
 * [Environment configuration](#environment-configuration)
   * [Switch between PHP versions](#switch-between-php-versions)
   * [Activating Varnish](#activating-varnish)
   * [Activating ElasticSearch](#activating-elasticsearch)
   * [Redis for caching](#redis-for-caching)
   * [Reset environment](#reset-environment)
   * [Switch NodeJS Versions](#switch-nodejs-versions)
 * [DevBox tests](#devbox-tests)
 * [FAQ](#faq)

## What You get

It's expected that the Magento 2 project source code will be located and managed on the host to allow quick indexing of project files by IDE. All other infrastructure is deployed on the guest machine.

Current DevBox aims to support multi-service deployment in one click. It is optimized for development scenarios using local environment.

The environment for Magento EE development is also configured.

<!--It is easy to [install multiple Magento instances](#multiple-magento-instances) based on different codebases simultaneously.-->

The [project initialization script](init_project.sh) configures a complete development environment:

<!-- 1. Adds some missing software on the host -->
 1. Configures all software necessary for Magento 2: Nginx, PHP 7.x, MySQL 5.6, Git, Composer, XDebug<!--, Rabbit MQ, Varnish-->
 1. Installs Magento 2 from Git repositories or Composer packages (can be configured via `checkout_source_from` option in [etc/config.yaml](etc/config.yaml.dist))
<!-- 1. Configures PHP Storm project (partially at the moment)-->
<!-- 1. Installs NodeJS, NPM, Grunt and Gulp for front end development

  :information_source: This box uses the [n package manager](https://www.npmjs.com/package/n) to provide the latest NodeJS LTS version.<br />
-->
## How to install

If you never used Kubernetes before, read the [Kubernetes Docs](https://kubernetes.io) first.

### Requirements

The software listed below should be available in [PATH](https://en.wikipedia.org/wiki/PATH_\(variable\)) (except for PHP Storm).

- [Docker](https://docs.docker.com/docker-for-mac/install/)
- [Minikube](https://kubernetes.io/docs/tasks/tools/install-minikube/)
- [Helm](https://docs.helm.sh/using_helm/#installing-helm)
- [VirtualBox](https://www.virtualbox.org/wiki/Downloads)
- [Git](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git) - Ensure that SSH keys are generated and associated with your Github account. See [how to check](https://help.github.com/articles/testing-your-ssh-connection/) and [how to configure](https://help.github.com/articles/generating-ssh-keys/), if not configured.<br />
  :information_source: To obtain the codebase without cloning, just use the Magento 2 codebase instead of `devbox-magento/magento2ce`. Either method will produce a successful installation.<br />

<!--  :information_source: On Windows hosts ![](docs/images/windows-icon.png) Git must be [v2.7+](http://git-scm.com/download/win). Also make sure to set the following options to avoid issues with incorrect line separators:

    ```
    git config --global core.autocrlf false
    git config --global core.eol LF
    git config --global diff.renamelimit 5000
    ```
-->
- [PHP Storm](https://www.jetbrains.com/phpstorm), optional but recommended. To get Helm support in PhpStorm make sure to get v2018.3+
- [NFS server](https://en.wikipedia.org/wiki/Network_File_System) ![](docs/images/linux-icon.png)![](docs/images/osx-icon.png) must be installed and running on \*nix and OSX hosts; usually available, follow [installation steps](#how-to-install) first

### Installation steps

:information_source: In case of any issues during installation, please read [FAQ section](#faq)

 1. Open terminal and change your directory to the one you want to contain Magento project. <!--![](docs/images/windows-icon.png) On Windows use Git Bash, which is available after Git installation.-->

 1. Download or clone the project with DevBox configuration:
    
    :warning: Do not open it in PhpStorm until `init_project.sh` has completed PhpStorm configuration in the initialize project step below.

     ```bash
     git clone --recursive git@github.com:magento/magento2-kubernetes-devbox.git magento2-devbox
     ```

    Optionally, if you use private repositories on GitHub or download packages from the Magento Marketplace using Composer.

     1. Copy [etc/composer/auth.json.dist](etc/composer/auth.json.dist) to `etc/composer/auth.json`.
     1. Specify your GitHub token by adding `"github.com": "your-github-token"` to the `github-oauth` section for GitHub authorization.
     1. Add the Magento Marketplace keys for Marketplace authorization to the `repo.magento.com` section.
     1. Copy (optional) [etc/config.yaml.dist](etc/config.yaml.dist) as `etc/config.yaml` and make the necessary customizations.

 1. Initialize the project (this will configure the environment, install Magento<!--, and configure the PHPStorm project-->):

    ```bash
    cd magento2-devbox
    
    # NFS configuration is needed just once for each project, it will prompt for your password to make changes on the host
    
    bash scripts/host/configure_nfs_exports.sh
    
    bash init_project.sh
    ```
    <!--
    To initialize project with checkout container,
    clone sources to checkout directory and use -e parameter to init_project.sh call.
    ```bash
    bash init_project.sh -e
    ```
    -->
    
 1. Use the `magento2-devbox` directory as the project root in PHP Storm (not `magento2-devbox/magento`). This is important, because in this case PHP Storm will be configured automatically by [init_project.sh](init_project.sh).<!-- If NFS files sync is disabled in [config](etc/config.yaml.dist) and ![](docs/images/windows-icon.png)on Windows hosts [verify the deployment configuration in PHP Storm](docs/phpstorm-configuration-windows-hosts.md).-->

    <!--Use the URL for accessing your Magento storefront in the browser as your Web server root URL. Typically this is the localhost, which refers to your development machine. Depending on how you've set up your VM you may also need a port number, like `http://localhost:8080`.-->

 1. Configure the remote PHP interpreter in PHP Storm. Go to Preferences, then Languages and Frameworks. Click PHP and add a new remote interpreter. Select Deployment configuration as a source for connection details.

### Default credentials and settings

Some of default settings are available for override. These settings can be found in the file [etc/config.yaml.dist](etc/config.yaml.dist).

To override settings create a copy of the file under the name 'config.yaml' and add your custom settings.

<!--When using [init_project.sh](init_project.sh), if not specified manually, random IP address is generated and is used as suffix for host name to prevent collisions, in case when two or more instances are running at the same time.-->

Upon a successful installation, you'll see the location and URL of the newly-installed Magento 2 application in console.

**Web access**:
- Access storefront at `http://<ip-from-config-yaml>`
- Access admin panel at `http://<ip-from-config-yaml>/admin/`
- Magento admin user/password: `admin/123123q`
- Rabbit MQ control panel: run `bash k-open-rabbitmq`, credentials `admin`/`123123q`

:information_source: Your admin URL, storefront URL, and admin user and password are located in `etc/config.yaml`.

**Codebase and DB access**:
- Path to your Magento installation in the container:
  - Can be retrieved from environment variable: `echo ${MAGENTO_ROOT}`
  <!--  - ![](docs/images/windows-icon.png) On Windows hosts: `/var/www/magento`-->
  - ![](docs/images/linux-icon.png)![](docs/images/osx-icon.png) On Mac and \*nix hosts: the same as on host
- MySQL DB host: 
  - inside the container: `localhost`
  - remotely: `<ip-from-config-yaml>` with port `30306`. Related configuration can be found in values.yaml inside magento2-devbox/etc/helm
- MySQL DB name: `magento`, `magento_integration_tests`
- MySQL DB user/password: `root:123123q`

**Codebase on host**
- CE codebase: `magento2-devbox/magento`
- EE codebase will be available if path to EE repository is specified in `etc/config.yaml`: `magento2-devbox/magento/magento2ee`

### Getting updates and fixes

Current devbox project follows [semantic versioning](http://semver.org/spec/v2.0.0.html) so feel free to pull the latest features and fixes, they will not break your project.
For example your current branch is `2.0`, then it will be safe to pull any changes from `origin/2.0`. However branch `3.0` will contain changes backward incompatible with `2.0`.
Note, that semantic versioning is only used for `x.0` branches (not for `develop` or `master`).

:information_source: To apply changes run `bash k-rebuild-environment`.

## Day-to-day development scenarios

### Access Magento

Use the following command to open current instance:
```bash
./m-open
```

Hostname can also be found in `maghento/host_name` section of [config.yaml](etc/config.yaml.dist).

### Reinstall Magento

Use commands described in [Switch between CE and EE](#switch-between-ce-and-ee) section with `-f` flag. Before doing actual re-installation, these commands update linking of EE codebase, clear cache, update composer dependencies.

If no composer update and relinking of EE codebase is necessary, use the following command. It will clear Magento DB, Magento caches and reinstall Magento instance.

Go to the root of the project in command line and execute:

```bash
./m-reinstall
```

### Clear Magento cache

Go to the root of the project in command line and execute:

```bash
./m-clear-cache
```

### Switch between CE and EE

Assume, that EE codebase is available in `magento2-devbox/magento/magento2ee`.
The following commands will link/unlink EE codebase, clear cache, update composer dependencies and reinstall Magento.
Go to 'magento2-devbox' created earlier and run in command line:

```bash
./m-switch-to-ce
# OR
./m-switch-to-ee
```

Force switch can be done using `-f` flag even if already switched to the target edition. May be helpful to relink EE modules after switching between branches.

Upgrade can be performed instead of re-installation using `-u` flag.

<!--:information_source: On Windows hosts (or when NFS mode is disabled in [config.yaml](etc/config.yaml.dist) explicitly) you will be asked to wait until code is uploaded to guest machine by PhpStorm (PhpStorm must be launched). To continue the process press any key.-->

### Sample data installation

Make sure that `ce_sample_data` and `ee_sample_data` are defined in [config.yaml](etc/config.yaml.dist) and point CE and optionally EE sample data repositories.
During initial project setup or during `bash init_project.sh -fc` (with `-fc` project will be re-created from scratch), sample data repositories willl be checked out to `magento2-devbox/magento/magento2ce-sample-data` and `magento2-devbox/magento/magento2ee-sample-data`.

To install Magento with sample data set `install_sample_data` in [config.yaml](etc/config.yaml.dist) to `1` and run `./m-switch-to-ce -f` or `./m-switch-to-ee -f`, depending on the edition to be installed. To disable sample data, set this option to `0` and force-switch to necessary edition (using the same commands).

### Basic data generation

Several entities are generated for testing purposes by default using REST API after Magento installation:
- Customer with address (credentials `customer@example.com`:`123123qQ`)
- Category
- Couple simple products
- Configurable product

To disable this feature, set `magento/generate_basic_data` in [config.yaml](etc/config.yaml.dist) to `0` and run `./m-switch-to-ce -f` or `./m-switch-to-ee -f`, depending on the edition to be installed.

### Use Magento CLI (bin/magento)

Go to 'magento2-devbox' created earlier and run in command line:

```bash
./m-bin-magento <command_name>
# e.g.
./m-bin-magento list
```

### Debugging with XDebug

XDebug is already configured to connect to the host machine automatically. So just:

 1. Set XDEBUG_SESSION=1 cookie (e.g. using 'easy Xdebug' extension for Firefox). See [XDebug documentation](http://xdebug.org/docs/remote) for more details
 1. Start listening for PHP Debug connections in PhpStorm on default 9000 port. See how to [integrate XDebug with PhpStorm](https://www.jetbrains.com/phpstorm/help/configuring-xdebug.html#integrationWithProduct)
 1. Set beakpoint or set option in PhpStorm menu 'Run -> Break at first line in PHP scripts'

<!--To debug a CLI script:

 1. Create [remote debug configuration](https://www.jetbrains.com/help/phpstorm/2016.1/run-debug-configuration-php-remote-debug.html) in PhpStorm, use `phpstorm` as IDE key
 1. Run created remote debug configuration
 1. Run CLI command on the guest as follows (`xdebug.remote_host` value might be different for you):

 ```bash
 php -d xdebug.remote_autostart=1 <path_to_cli_script>
 ```

To debug Magento Setup script, go to [Magento installation script](scripts/guest/m-reinstall) and find `php ${install_cmd}`. Follow steps above for any CLI script

:information_source: In addition to XDebug support, [config.yaml](etc/config.yaml.dist) has several options in `debug` section which allow storefront and admin UI debugging. Plus, desired Magento mode (developer/production/default) can be enabled using `magento_mode` option, default is developer mode.
-->
### Connecting to MySQL DB

Go to 'magento2-devbox' created earlier and run in command line:

```bash
bash k-ssh-mysql
```

After successful login to the container run the following command and enter `123123q` when prompted for a password:

```bash
mysql -uroot -p
```

### View emails sent by Magento

Not available yet.
<!--All emails are saved to 'magento2-devbox/log/email' in HTML format.-->

### Accessing PHP and other config files

The following configuration files are used by default:
- [NGINX](etc/helm/templates/configmap.yaml)
- PHP-FPM: [ini](scripts/etc/php-fpm.ini), [conf](scripts/etc/php-fpm.conf)
- [xDebug](scripts/etc/php-xdebug.ini)
- [Dockerfile for monolith base image](scripts/Dockerfile)
- [Actually applied Dockerfile for monolith with customizations](etc/docker/monolith/Dockerfile)
- [Actually applied Dockerfile for monolith with xDebug and customizations](etc/docker/monolith-with-xdebug/Dockerfile)
- [Kubernetes config for Monolith](etc/helm/templates/magento2-deployment.yaml)
- [Kubernetes Helm variables](etc/helm/values.yaml)
<!--- [Configs for Checkout service](etc/helm/charts/checkout)-->
<!--It is possible to view/modify majority of guest machine config files directly from IDE on the host. They will be accessible in [etc/guest](etc/guest) directory only when guest machine is running. The list of accessible configs includes: PHP, Apache, Mysql, Varnish, RabbitMQ.
Do not edit any symlinks using PhpStorm because it may break your installation.

After editing configs in IDE it is still required to restart related services manually.
-->
### Upgrading Magento

Sometimes it is necessary to test upgrade flow. This can be easily done as follows (assuming that you have installed instance):

 - For git-based installation - check out codebase corresponding to the target Magento version. Or modify your `composer.json` in case of composer-based installation
 - Use commands described in [Switch between CE and EE](#switch-between-ce-and-ee) section with `-u` flag

### Multiple Magento instances

Not available yet.
<!--To install several Magento instances based on different code bases, just follow [Installation steps](#installation-steps) to initialize project in another directory on the host.
Unique IP address, SSH port and domain name will be generated for each new instance if not specified manually in `etc/config.yaml`
-->
### Update Composer dependencies

Go to 'magento2-devbox' created earlier and run in command line:

```bash
./m-composer install
# OR
./m-composer update
```

### Running Magento tests

Not available yet.
<!--See [draft](https://github.com/paliarush/magento2-vagrant-for-developers/issues/120)-->


## Environment configuration

### Switch between PHP versions

Not available yet.
<!--Switch between PHP versions using "php_version: <version>" option in [config.yaml](etc/config.yaml.dist). Supported versions are 5.6, 7.0, 7.1 and 7.2.
PHP version will be applied after "devbox reload".
-->

### Activating Varnish

Not available yet.
<!--Set `use_varnish: 1` to use varnish along with apache in [config.yaml](etc/config.yaml.dist). Changes will be applied on `m-reinstall`.
It will use default file etc/magento2_default_varnish.vcl.dist generated from a Magento 2.1 instance. Magento 2.2+ supports additional Varnish features and you may need to provide custom version of VCL to enable them.
Varnish Version: 4.1

Use the following commands to enable/disable varnish without reinstalling Magento: `m-varnish disable` or `m-varnish enable`.
-->

### Activating ElasticSearch

Set `search_engine: "elasticsearch"` in [config.yaml](etc/config.yaml.dist) to use ElasticSearch as current search engine or `search_engine: "mysql"` to use MySQL. Changes will be applied on `m-reinstall`.

Use the following commands to switch between search engines without reinstalling Magento: `m-search-engine elasticsearch` or `m-search-engine mysql`.

### Redis for caching

<!--:information_source: Available in Magento v2.0.6 and higher.-->

Redis is configured as cache backend by default. <!--It is still possible to switch back to filesystem cache by changing `environment_cache_backend` to `filesystem` in [config.yaml](etc/config.yaml.dist).-->

### Reset environment

It is possible to reset project environment to default state, which you usually get just after project initialization. The following command will re-initialize Kubernetes cluster. Magento 2 code base (`magento` directory) and [etc/config.yaml](etc/config.yaml.dist) and PhpStorm settings will stay untouched<!--, but guest config files (located in [etc/guest](etc/guest)) will be cleared-->.

Go to 'magento2-devbox' created earlier and run in command line:

```bash
./init_project.sh -f
```

It is possible to reset Magento 2 code base at the same time. Magento 2 code base will be deleted and then cloned from the repositories specified in [etc/config.yaml](etc/config.yaml.dist)

```bash
./init_project.sh -fc
```

To reset PhpStorm project configuration, in addition to `-f` specify `-p` option:

```bash
./init_project.sh -fp
```

Ultimate project reset can be achieved by combining all available flags:

```bash
./init_project.sh -fcp
```

### Switch NodeJS Versions

NodeJS not available yet.
<!--
By default, the box will install the latest `NodeJS LTS` version using the [n package manager](https://www.npmjs.com/package/n). If you need another version of `Node` because of Magento's `package.json` requirements, simply run:

```js
n <version>
```

Note: See [Working with npm](https://www.npmjs.com/package/n#working-with-npm) if after switching versions with `n`, `npm` is not working properly.
-->

### DevBox tests

The tests are executed on every PR on Travis CI. It is possible to configure the same tests to run on the forked repository.
In order to run composer-based Magento tests for the fork, repo.magento.com credentials must be set to `COMPOSER_AUTH` [environment variable](https://docs.travis-ci.com/user/environment-variables/#defining-variables-in-repository-settings) on Travis CI, the variable value should be:
```json
'{"http-basic": {"repo.magento.com": {"username": "<public_key>","password": "<secret_key>"}}}'
```

An extended testsuite by default is executed against the master branch only.
It is possible to execute an extended testsuite on every build by commenting out `if: branch = master` in the [.travis.yaml](./.travis.yml)

The same tests can be run on local using the following command. :warning: only one devbox can be running on the same host at the same time. The tests will destroy existing devbox installation.

```bash
cd tests
bash ./<test-name>.sh
```

### FAQ
 1. To debug any CLI script in current Devbox project, set `debug:devbox_project` option in [config.yaml](etc/config.yaml.dist) to `1`
 1. Make sure that you used `magento2-devbox` directory as project root in PHP Storm (not `magento2-devbox/magento`)
 1. If project opened in PhpStorm looks broken, close PhpStorm  and remove `magento2-devbox/.idea`. Run `./magento2-devbox/scripts/host/configure_php_storm.sh`. After opening project in PhpStorm again everything should look good
 1. Please make sure that currently installed software, specified in [requirements section](#requirements), meets minimum version requirement
 1. Be careful if your OS is case-insensitive, NFS might break the symlinks if you cd into the wrong casing and you power the devbox up. Just be sure to cd in to the casing the directory was originally created as.
 1. Cannot run unit tests from PHPStorm on Magento 2.2, see possible solution [here](https://github.com/paliarush/magento2-vagrant-for-developers/issues/167)
 1. [Permission denied (publickey)](https://github.com/paliarush/magento2-vagrant-for-developers/issues/165)


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

 * [\[Unreleased\]](#unreleased)

## [Unreleased]

### Changed

### Added

### Fixed


init_project.sh
#!/usr/bin/env bash

## TODO: Fix
set -e

devbox_dir=$PWD

source "${devbox_dir}/scripts/functions.sh"
resetNestingLevel
current_script_name=`basename "$0"`
initLogFile ${current_script_name}

debug_devbox_project="$(bash "${devbox_dir}/scripts/get_config_value.sh" "debug_devbox_project")"
if [[ ${debug_devbox_project} -eq 1 ]]; then
    set -x
fi

config_path="${devbox_dir}/etc/config.yaml"
if [[ ! -f "${config_path}" ]]; then
    status "Initializing etc/config.yaml using defaults from etc/config.yaml.dist"
    cp "${config_path}.dist" "${config_path}"
fi

magento_ce_dir="${devbox_dir}/magento"
magento_ce_sample_data_dir="${magento_ce_dir}/magento2ce-sample-data"
magento_ee_dir="${magento_ce_dir}/magento2ee"
magento_ee_sample_data_dir="${magento_ce_dir}/magento2ee-sample-data"
host_os="$(bash "${devbox_dir}/scripts/host/get_host_os.sh")"
use_nfs="$(bash "${devbox_dir}/scripts/get_config_value.sh" "guest_use_nfs")"
nfs_server_ip="$(bash "${devbox_dir}/scripts/get_config_value.sh" "guest_nfs_server_ip")"
repository_url_ce="$(bash "${devbox_dir}/scripts/get_config_value.sh" "repository_url_ce")"
#repository_url_checkout="$(bash "${devbox_dir}/scripts/get_config_value.sh" "repository_url_checkout")"
repository_url_ee="$(bash "${devbox_dir}/scripts/get_config_value.sh" "repository_url_ee")"
composer_project_name="$(bash "${devbox_dir}/scripts/get_config_value.sh" "composer_project_name")"
composer_project_url="$(bash "${devbox_dir}/scripts/get_config_value.sh" "composer_project_url")"
checkout_source_from="$(bash "${devbox_dir}/scripts/get_config_value.sh" "checkout_source_from")"

function checkoutSourceCodeFromGit()
{
    if [[ ! -d ${magento_ce_dir} ]]; then
#        if [[ ${host_os} == "Windows" ]]; then
#            status "Configuring git for Windows host"
#            git config --global core.autocrlf false
#            git config --global core.eol LF
#            git config --global diff.renamelimit 5000
#        fi

        initMagentoCeGit
        initMagentoCeSampleGit

        # By default EE repository is not specified and EE project is not checked out
        if [[ -n "${repository_url_ee}" ]]; then
            initMagentoEeGit
        fi
        # By default EE sample data repository is not specified and EE project is not checked out
        repository_url_ee_sample_data="$(bash "${devbox_dir}/scripts/get_config_value.sh" "repository_url_ee_sample_data")"
        if [ -n "${repository_url_ee_sample_data}" ]; then
            initMagentoEeSampleGit
        fi
    fi
}

function initMagentoCeGit()
{
    initGitRepository ${repository_url_ce} "CE" "${magento_ce_dir}"
}

function initMagentoEeGit()
{
    initGitRepository ${repository_url_ee} "EE" "${magento_ee_dir}"
}

function initMagentoCeSampleGit()
{
    repository_url_ce_sample_data="$(bash "${devbox_dir}/scripts/get_config_value.sh" "repository_url_ce_sample_data")"
    initGitRepository ${repository_url_ce_sample_data} "CE sample data" "${magento_ce_sample_data_dir}"
}

function initMagentoEeSampleGit()
{
    initGitRepository ${repository_url_ee_sample_data} "EE sample data" "${magento_ee_sample_data_dir}"
}

# Initialize the cloning and checkout of a git repository
# Arguments:
#   Url of repository
#   Name of repository (CE, EE)
#   Directory where the repository will be cloned to
function initGitRepository()
{
    local repository_url=${1}
    local repository_name=${2}
    local directory=${3}

    if [[ ${repository_url} == *"::"* ]]; then
        local branch=$(getGitBranch ${repository_url})
        local repo=$(getGitRepository ${repository_url})
    else
        local repo=${repository_url}
    fi

    status "Checking out ${repository_name} repository"
    git clone ${repo} "${directory}" 2> >(logError) > >(log)

    if [[ -n ${branch} ]]; then
        status "Checking out branch ${branch} of ${repository_name} repository"
        cd "${directory}"
        git fetch 2> >(logError) > >(log)
        git checkout ${branch} 2> >(log) > >(log)
    fi
    cd "${devbox_dir}"
}

# Get the git repository from a repository_url setting in config.yaml
function getGitRepository()
{
    local repo="${1%::*}" # Gets the substring before the '::' characters
    echo ${repo}
}

# Get the git branch from a repository_url setting in config.yaml
function getGitBranch()
{
    local branch="${1#*::}" # Gets the substring after the '::' characters
    echo ${branch}
}

function composerCreateProject()
{
    if [[ ! -d ${magento_ce_dir} ]]; then
        status "Downloading Magento codebase using 'composer create-project'"
        bash "${devbox_dir}/scripts/host/m_composer.sh" create-project ${composer_project_name} "${magento_ce_dir}" --repository-url=${composer_project_url}

#        # TODO: Workaround for Magento 2.2+ until PHP is upgraded to 7.1 on the guest
#        cd "${magento_ce_dir}"
#        composer_dir="${devbox_dir}/scripts/host"
#        composer_phar="${composer_dir}/composer.phar"
#        php_executable="$(bash "${devbox_dir}/scripts/host/get_path_to_php.sh")"
#        project_version="$("${php_executable}" "${composer_phar}" show --self | grep version)"
#        matching_version_pattern='2.[23].[0-9]+'
#        if [[ ${project_version} =~ ${matching_version_pattern} ]]; then
#            status "Composer require zendframework/zend-code:~3.1.0 (needed for Magento 2.2+ only)"
#            cd "${magento_ce_dir}"
#            bash "${devbox_dir}/scripts/host/m_composer.sh" require "zendframework/zend-code:~3.1.0"
#        fi
    fi
}

bash "${devbox_dir}/scripts/host/check_requirements.sh"

# Clean up the project before initialization if "-f" option was specified. Remove codebase if "-fc" is used.
force_project_cleaning=0
force_codebase_cleaning=0
force_phpstorm_config_cleaning=0
enable_checkout=0
while getopts 'fcpe' flag; do
  case "${flag}" in
    f) force_project_cleaning=1 ;;
    c) force_codebase_cleaning=1 ;;
    p) force_phpstorm_config_cleaning=1 ;;
    e) enable_checkout=1 ;;
    *) error "Unexpected option" && exit 1;;
  esac
done
if [[ ${force_project_cleaning} -eq 1 ]]; then
    status "Cleaning up the project before initialization since '-f' option was used"

# TODO: Remove if not needed in the future
#    if [[ $(isMinikubeInitialized) -eq 1 ]]; then
#        bash "${devbox_dir}/scripts/host/k_rebuild_environment.sh"
#    fi


    if [[ $(isMinikubeRunning) -eq 1 ]]; then
        minikube stop 2> >(logError) | {
          while IFS= read -r line
          do
            filterDevboxOutput "${line}"
            lastline="${line}"
          done
          filterDevboxOutput "${lastline}"
        }
    fi
    if [[ $(isMinikubeStopped) -eq 1 ]]; then
        minikube delete 2> >(logError) | {
          while IFS= read -r line
          do
            filterDevboxOutput "${line}"
            lastline="${line}"
          done
          filterDevboxOutput "${lastline}"
        }
    fi

    mv "${devbox_dir}/etc/guest/.gitignore" "${devbox_dir}/etc/.gitignore.back"
    rm -rf "${devbox_dir}/.devbox" "${devbox_dir}/etc/guest"
    mkdir "${devbox_dir}/etc/guest"
    mv "${devbox_dir}/etc/.gitignore.back" "${devbox_dir}/etc/guest/.gitignore"
    cd "${devbox_dir}/log" && mv email/.gitignore email_gitignore.back && rm -rf email && mkdir email && mv email_gitignore.back email/.gitignore
    if [[ ${force_codebase_cleaning} -eq 1 ]]; then
        status "Removing current Magento codebase before initialization since '-c' option was used"
        rm -rf "${magento_ce_dir}"
    fi
fi

status "Initializing dev box"
cd "${devbox_dir}"

#if [[ $(isMinikubeInitialized) -eq 1 ]]; then
#    warning "The project has already been initialized.
#    To re-initialize the project add the '-f' flag (using just '-f' will not affect Magento codebase or PHP Storm settings).
#    To delete Magento codebase and initialize it from scratch based on etc/config.yaml add '-c' flag.
#    To reconfigure PHP Storm add '-p' flag."
#    exit 0
#fi

if [[ $(isMinikubeRunning) -eq 0 ]]; then
    status "Starting minikube"
    minikube start --cpus=2 --memory=4096
    minikube addons enable ingress
    minikube addons enable heapster
    # hanged in some cases todo
#    minikube start --cache-images --cpus=2 --memory=4096 2> >(logError) | {
#      while IFS= read -r line
#      do
#        filterDevboxOutput "${line}"
#        lastline="${line}"
#      done
#      filterDevboxOutput "${lastline}"
#    }
fi

config_content="$(cat ${config_path})"
default_nfs_server_ip_pattern="nfs_server_ip: \"0\.0\.0\.0\""
if [[ ! ${config_content} =~ ${default_nfs_server_ip_pattern} ]]; then
    status "Custom NFS server IP is already specified in etc/config.yaml (${nfs_server_ip})"
else
    nfs_server_ip="$(minikube ip | grep -oh ^[0-9]*\.[0-9]*\.[0-9]*\. | head -1 | awk '{print $1"1"}')"
    status "Saving NFS server IP to etc/config.yaml (${nfs_server_ip})"
    sed -i.back "s|${default_nfs_server_ip_pattern}|nfs_server_ip: \"${nfs_server_ip}\"|g" "${config_path}"
    rm -f "${config_path}.back"
fi

status "Configuring kubernetes cluster on the minikube"
# TODO: Optimize. Helm tiller must be initialized and started before environment configuration can begin
helm init --wait
#waitForKubernetesPodToRun 'tiller-deploy'

# TODO: change k-rebuild-environment to comply with formatting requirements

if [[ "${enable_checkout}" == 1 ]]; then
    bash "${devbox_dir}/scripts/host/k_rebuild_environment.sh" -e
else
    bash "${devbox_dir}/scripts/host/k_rebuild_environment.sh"
fi

if [[ ! -d ${magento_ce_dir} ]]; then
    if [[ "${checkout_source_from}" == "composer" ]]; then
        composerCreateProject
    elif [[ "${checkout_source_from}" == "git" ]]; then
        checkoutSourceCodeFromGit
    else
        error "Value specified for 'checkout_source_from' is invalid. Supported options: composer OR git"
        exit 1
    fi
fi

monolith_ip="$(minikube service magento2-monolith --url | grep -oE '[0-9][^:]+' | head -1)"
status "Saving Magento monolith container IP to etc/config.yaml (${monolith_ip})"
sed -i.back "s|ip_address: \".*\"|ip_address: \"${monolith_ip}\"|g" "${config_path}"
sed -i.back "s|host_name: \".*\"|host_name: \"${monolith_ip}\"|g" "${config_path}"
rm -f "${config_path}.back"

bash "${devbox_dir}/scripts/host/check_mounted_directories.sh"

if [[ ${force_project_cleaning} -eq 1 ]] && [[ ${force_phpstorm_config_cleaning} -eq 1 ]]; then
    status "Resetting PhpStorm configuration since '-p' option was used"
    rm -rf "${devbox_dir}/.idea"
fi
if [[ ! -f "${devbox_dir}/.idea/deployment.xml" ]]; then
    bash "${devbox_dir}/scripts/host/configure_php_storm.sh"
fi
bash "${devbox_dir}/scripts/host/configure_tests.sh"

#if [[ ${host_os} == "Windows" ]] || [[ ${use_nfs} == 0 ]]; then
#    # Automatic switch to EE during project initialization cannot be supported on Windows
#    status "Installing Magento CE"
#    bash "${devbox_dir}/scripts/host/m_reinstall.sh" 2> >(logError)
#else
    if [[ -n "${repository_url_ee}" ]]; then
        bash "${devbox_dir}/scripts/host/m_switch_to_ee.sh" -f 2> >(logError)
    else
        bash "${devbox_dir}/scripts/host/m_switch_to_ce.sh" -f 2> >(logError)
    fi
#fi

success "Project initialization succesfully completed (make sure there are no errors in the log above)"

info "$(bold)[Important]$(regular)
    Please use $(bold)${devbox_dir}$(regular) directory as PhpStorm project root, NOT $(bold)${magento_ce_dir}$(regular)."

#if [[ ${host_os} == "Windows" ]] || [[ ${use_nfs} == 0 ]]; then
#    info "$(bold)[Optional]$(regular)
#    To verify that deployment configuration for $(bold)${magento_ce_dir}$(regular) in PhpStorm is correct,
#        use instructions provided here: $(bold)https://github.com/paliarush/magento2-vagrant-for-developers/blob/2.0/docs/phpstorm-configuration-windows-hosts.md$(regular).
#    If not using PhpStorm, you can set up synchronization using rsync"
#fi

info "$(regular)See details in $(bold)${devbox_dir}/log/${current_script_name}.log$(regular). For debug output set $(bold)debug:devbox_project$(regular) to $(bold)1$(regular) in $(bold)etc/config.yaml$(regular)"
